**GI domaći zadatak 2**

In [1]:
#docker run -p 8888:8888 jupyter/minimal-notebook 
!pip install pysam

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.6/15.6 MB 1.2 MB/s eta 0:00:0000:0100:01m


In [2]:
import pysam

In [3]:
save = pysam.set_verbosity(0)
bamfile=pysam.AlignmentFile("merged-tumor.bam","rb")
pysam.set_verbosity(save)

0

In [4]:
j=1
for read in bamfile:
    print(f'Informacije o ',j,'. read-u:')
    print(read)
    print(f'\n')
    print(f'QNAME ',j,':',read.qname)
    print(f'FLAG ',j,':',read.flag)
    print(f'RNAME ',j,':',read.rname)
    print(f'POS ',j,':', read.pos)
    print(f'MAPQ ',j,':',read.mapq)
    print(f'CIGAR ',j,':',read.cigarstring)
    print(f'MRNM ',j,':',read.mrnm)
    print(f'MPOS ',j,':', read.mpos)
    print(f'ISIZE ',j,':', read.isize)
    print(f'SEQ ',j,':', read.seq)
    print(f'QUAL ',j,':', read.qual)
    print(f'TAGS ',j,':', read.get_tags())
    print(f'\nInformacije o flegu read-a ',j,':')
    print(f'Uparen read u sekvenciranju: ',read.is_paired)
    print(f'Mapiran read u svom paru: ',read.is_proper_pair)
    print(f'Nemapiran read: ',read.is_unmapped)
    print(f'Nemapiran par read-a: ',read.mate_is_unmapped)
    print(f'Reverse/forward strand: ',read.is_reverse)
    print(f'Reverse/forward strand para: ',read.mate_is_reverse)
    print(f'Read je prvi read u paru: ',read.is_read1)
    print(f'Read je drugi read u paru: ',read.is_read2)
    print(f'Poravnanje nije primarno:', read.is_secondary)
    print(f'Read nije uspešno proverio kvalitet:', read.is_qcfail)
    print(f'Read je ili PCR ili optički duplikat:' , read.is_duplicate)
    j=j+1
    if (j==5):
        break
    print(f'\n')
#bamfile.close()

Informacije o  1 . read-u:
C0HVYACXX120402:7:1207:5722:57044	1187	#20	9483249	27	76M	#20	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]


QNAME  1 : C0HVYACXX120402:7:1207:5722:57044
FLAG  1 : 1187
RNAME  1 : 20
POS  1 : 9483248
MAPQ  1 : 27
CIGAR  1 : 76M
MRNM  1 : 20
MPOS  1 : 9483381
ISIZE  1 : 209
SEQ  1 : TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAA

**Informacije o  1 . read-u:**
C0HVYACXX120402:7:1207:5722:57044	1187	#20	9483249	27	76M	#20	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 28, 27, 29, 31, 30, 31, 31, 29, 31, 35, 30, 29, 31, 34, 30, 29, 23, 41, 32, 20, 30, 29, 34, 34, 29, 30, 31, 30, 30, 30, 33, 33, 26, 39, 12, 25, 19, 32, 30, 35, 28, 35, 33, 23, 33, 35, 36, 30, 38, 33, 41, 34, 35, 31, 33, 23, 30, 30, 36, 27, 32, 29, 34, 35, 41, 33, 31, 33, 29, 32, 32, 31, 31, 31, 34])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]


**QNAME  1 :** C0HVYACXX120402:7:1207:5722:57044
**FLAG  1 :** 1187
**RNAME  1 :** 20
**POS  1 :** 9483248
**MAPQ  1 :** 27
**CIGAR  1 :** 76M
**MRNM  1 :** 20
**MPOS  1 :** 9483381
**ISIZE  1 :** 209
**SEQ  1 :** TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG
**QUAL  1 :** ==<>@?@@>@D?>@C?>8JA5?>CC>?@???BB;H-:4A?D=DB8BDE?GBJCD@B8??E<A>CDJB@B>AA@@@C
**TAGS  1 :** [('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]

**Informacije o flegu read-a  1 :**
**Uparen read u sekvenciranju:** True
**Mapiran read u svom paru:**  True
**Nemapiran read:**  False
**Nemapiran par read-a:**  False
**Reverse/forward strand:**  False
**Reverse/forward strand para:**  True
**Read je prvi read u paru:**  False
**Read je drugi read u paru:**  True
**Poravnanje nije primarno:** False
**Read nije uspešno proverio kvalitet:** False
**Read je ili PCR ili optički duplikat:** True


**Informacije o  2 . read-u:**
D0RE2ACXX120401:1:2105:2631:113383	163	#20	9483249	27	76M	#20	9483382	209	TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG	array('B', [28, 29, 27, 30, 31, 30, 31, 31, 34, 31, 40, 30, 29, 34, 41, 34, 29, 33, 39, 40, 37, 36, 30, 39, 36, 34, 29, 35, 26, 41, 31, 33, 31, 35, 39, 34, 34, 29, 34, 31, 36, 37, 35, 35, 33, 35, 36, 35, 40, 37, 34, 39, 30, 29, 28, 15, 19, 30, 36, 36, 36, 33, 33, 38, 34, 37, 33, 30, 33, 33, 29, 34, 32, 30, 31, 29])	[('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]


**QNAME  2 :** D0RE2ACXX120401:1:2105:2631:113383
**FLAG  2 :** 163
**RNAME  2 :** 20
**POS  2 :** 9483248
**MAPQ  2 :** 27
**CIGAR  2 :** 76M
**MRNM  2 :** 20
**MPOS  2 :** 9483381
**ISIZE  2 :** 209
**SEQ  2 :** TTTTCAAACAGTATCTATGCCTGCCAAATGTGAACATATAAAAAAAAACCAGAATGTGCCATTCTGATTTAAACTG
**QUAL  2 :** =><?@?@@C@I?>CJC>BHIFE?HEC>D;J@B@DHCC>C@EFDDBDEDIFCH?>=04?EEEBBGCFB?BB>CA?@>
**TAGS  2 :** [('XA', 'GL000217.1,-110754,76M,1;'), ('BD', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('MD', '76'), ('RG', '1'), ('BI', 'NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN'), ('NM', 0), ('MQ', 27), ('AS', 76), ('XS', 71)]

**Informacije o flegu read-a  2 :**
**Uparen read u sekvenciranju:**  True
**Mapiran read u svom paru:**  True
**Nemapiran read:**  False
**Nemapiran par read-a:**  False
**Reverse/forward strand:**  False
**Reverse/forward strand para:**  True
**Read je prvi read u paru:**  False
**Read je drugi read u paru:**  True
**Poravnanje nije primarno:** False
**Read nije uspešno proverio kvalitet:** False
**Read je ili PCR ili optički duplikat:** False


In [5]:
unmapnum=0
for read in bamfile.fetch():
    if read.is_unmapped:
        unmapnum=unmapnum+1
print(f'Ukupan broj nemapiranih read-ova',unmapnum)

Ukupan broj nemapiranih read-ova 17765


Ukupan broj nemapiranih read-ova **17765**

In [6]:
num=bamfile.count()
print(f'Ukupan broj read-ova:',num)

Ukupan broj read-ova: 2921629


Ukupan broj read-ova: **2921629**

In [7]:
zeroqual_num=0
for read in bamfile.fetch():
    if read.mapq==0:
        zeroqual_num=zeroqual_num+1
print(f'Broj read-ova sa kvalitetom mapiranja 0:',zeroqual_num)

Broj read-ova sa kvalitetom mapiranja 0: 126628


Broj read-ova sa kvalitetom mapiranja 0: **126628**

In [9]:
sum=0
for read in bamfile.fetch():
    sum=sum+read.mapq
print(f'Prosečni kvalitet mapiranja svih read-ova:',sum/num)

Prosečni kvalitet mapiranja svih read-ova: 55.91379158681681


Prosečni kvalitet mapiranja svih read-ova: **55.91379158681681**

In [10]:
#prvi nacin
sum2=0
numNonZero=0
for read in bamfile.fetch():
    if read.mapq!=0:
        sum2=sum2+read.mapq
        numNonZero=numNonZero+1
print(f'Prosečni kvalitet mapiranja svih read-ova sa kvalitetom različitim od 0:',sum2/numNonZero)

Prosečni kvalitet mapiranja svih read-ova sa kvalitetom različitim od 0: 58.446975510921106


Prosečni kvalitet mapiranja svih read-ova sa kvalitetom različitim od 0 dobijen na prvi način: **58.446975510921106**

In [12]:
#drugi nacin
print(f'Prosečni kvalitet mapiranja svih read-ova sa kvalitetom različitim od 0:',sum/(num-zeroqual_num))

Prosečni kvalitet mapiranja svih read-ova sa kvalitetom različitim od 0: 58.446975510921106


Prosečni kvalitet mapiranja svih read-ova sa kvalitetom različitim od 0 dobijen na drugi način: **58.446975510921106**